dmoj/cptbox/utils.py

```diff
@@ -1,11 +1,62 @@
+import errno
 import io
+import mmap
+import os
+from tempfile import NamedTemporaryFile
+from typing import Optional
 
 from dmoj.cptbox._cptbox import memory_fd_create, memory_fd_seal
+from dmoj.cptbox.tracer import FREEBSD
 
 
 class MemoryIO(io.FileIO):
-    def __init__(self) -> None:
-        super().__init__(memory_fd_create(), 'r+')
+    _name: Optional[str] = None
+
+    def __init__(self, prefill: Optional[bytes] = None, seal=False) -> None:
+        if FREEBSD:
+            with NamedTemporaryFile(delete=False) as f:
+                self._name = f.name
+                super().__init__(os.dup(f.fileno()), 'r+')
+        else:
+            super().__init__(memory_fd_create(), 'r+')
+
+        if prefill:
+            self.write(prefill)
+        if seal:
+            self.seal()
 
     def seal(self) -> None:
-        memory_fd_seal(self.fileno())
+        fd = self.fileno()
+        try:
+            memory_fd_seal(fd)
+        except OSError as e:
+            if e.errno == errno.ENOSYS:
+                # FreeBSD
+                self.seek(0, os.SEEK_SET)
+                return
+            raise
+
+        new_fd = os.open(f'/proc/self/fd/{fd}', os.O_RDONLY)
+        try:
+            os.dup2(new_fd, fd)
+        finally:
+            os.close(new_fd)
```

Comment:
Maybe we could use a comment about why this dup is needed. Also, why isn't it implemented in the C function?

Reply:
C code is just that much more painful to maintain.

Reply:
Factored it to a function with an obvious name.